control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go

```diff
@@ -143,6 +146,14 @@ func (r *reconciler) reconcile(ctx context.Context) error {
 		return fmt.Errorf("failed to get hosted control plane %s/%s: %w", r.hcpNamespace, r.hcpName, err)
 	}
 
+	expectedImage, err := controllers.GetExpectedOperatorImage(ctx, r.cpClient, r.releaseProvider, hcp)
+	if err != nil {
+		return fmt.Errorf("failed to get expected image: %w", err)
+	}
+	if expectedImage != r.activeImage {
+		log.Info("No-oping until recreated with expected image", "activeImage", r.activeImage, "expectedImage", expectedImage)
```

Comment:
Is it safe to abruptly short-circuit like this? i.e wouldn't we need to check the HCP.Status as well to validate it's stable and no upgrades in-place so it's safe to short-circuit?

Reply:
yep and is necessary to short circuit. Same explanation as above: but the only caveat is the delete path needs to still be allowed because at that point higher level components do not reconcile anymore. If it wasn't allowed it would be a deadlock.